server/etcdserver/api/v3discovery/discovery.go

```diff
@@ -0,0 +1,544 @@
+	}
+	return ErrTooManyRetries
+}
+
+func (d *discovery) registerSelf(contents string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), d.cfg.RequestTimeOut)
+	memberKey := getMemberKey(d.cluster, d.id.String())
+	_, err := d.c.Put(ctx, memberKey, contents)
+	cancel()
+
+	if err != nil {
+		d.lg.Warn(
+			"failed to register members itself to the discovery service",
+			zap.String("discovery-url", d.durl),
+			zap.String("memberKey", memberKey),
+			zap.Error(err),
+		)
+		return d.registerSelfRetry(contents)
+	}
+
+	d.lg.Info(
+		"register member itself successfully",
+		zap.String("discovery-url", d.durl),
+		zap.String("memberKey", memberKey),
+		zap.String("memberInfo", contents),
+	)
+
+	return nil
+}
+
+func (d *discovery) waitNodes(members memberInfoMap, clusterSize int, rev int64) (memberInfoMap, error) {
+	// watch from the next revision
+	membersKeyPrefix := getMemberKeyPrefix(d.cluster)
+	w := d.c.Watch(context.Background(), membersKeyPrefix, clientv3.WithPrefix(), clientv3.WithRev(rev+1))
+
+	d.lg.Info(
+		"waiting for peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+
+	// waiting for peers until all needed peers are returned
+	for wresp := range w {
+		for _, ev := range wresp.Events {
+			mKey := strings.TrimSpace(string(ev.Kv.Key))
+			mValue := strings.TrimSpace(string(ev.Kv.Value))
+
+			if shouldIgnorePeer(members, mKey, mValue, d.cluster, d.lg, d.durl) {
+				continue
+			}
+
+			d.lg.Info(
+				"found peer from discovery service",
+				zap.String("discovery-url", d.durl),
+				zap.String("memberKey", mKey),
+				zap.String("memberInfo", mValue),
+			)
+
+			members[mKey] = memberInfo{peerURL: mValue, createRev: ev.Kv.CreateRevision}
+		}
+
+		if len(members) >= clusterSize {
+			break
+		}
+	}
+
+	d.lg.Info(
+		"found all needed peers from discovery service",
+		zap.String("discovery-url", d.durl),
+		zap.Int("clusterSize", clusterSize),
+		zap.Int("found-peers", len(members)),
+	)
+	return members, nil
+}
+
+func shouldIgnorePeer(members memberInfoMap, memberKey, memberInfo string, cluster string, lg *zap.Logger, durl string) bool {
+	membersKeyPrefix := getMemberKeyPrefix(cluster)
+
+	if !strings.HasPrefix(memberKey, membersKeyPrefix) {
+		// It should never happen.
+		lg.Panic(
+			"invalid peer registry key",
+			zap.String("discovery-url", durl),
+			zap.String("memberKey", memberKey),
+			zap.String("memberInfo", memberInfo),
+		)
+	}
+
+	if strings.IndexRune(memberInfo, '=') == -1 {
+		// It must be in the format "member1=http://127.0.0.1:2380".
+		lg.Warn(
+			"invalid peer info returned from discovery service",
+			zap.String("discovery-url", durl),
+			zap.String("memberKey", memberKey),
+			zap.String("memberInfo", memberInfo),
+		)
+		return true
+	}
+
+	if _, ok := members[memberKey]; ok {
+		lg.Warn(
+			"found duplicate peer from discovery service",
+			zap.String("discovery-url", durl),
```

Comment:
durl argument is just passed so it can be added for logger. We should not pass values like that rather we could zap feature to mark whole logger.

For example:
```
func caller(lg *zap.Logger, durl string) {
  doA(lg, durl)
  doB(lg, durl)
}

func doA(lg *zap.Logger, durl string) {
  lg.Info("did A", zap.String('discovery-url", durl))
}
func doB(lg *zap.Logger, durl string) {
  lg.Info("did B", zap.String('discovery-url", durl))
}
```
Should be replaced with
```
func caller(lg *zap.Logger, durl string) {
  lg = lg.With(zap.String('discovery-url", durl))
  doA(lg)
  doB(lg)
}

func doA(lg *zap.Logger) {
  lg.Info("did A")
}
func doB(lg *zap.Logger, durl string) {
  lg.Info("did B")
}
```

Reply:
Makes sense, thanks & resolved.